src/system/sargon_os/sargon_os_factors.rs

```diff
@@ -1,7 +1,164 @@
+    }
+
+    /// Updates the factor source `updated` by mutating current profile and persisting
+    /// the change to secure storage. Throws `UpdateFactorSourceMutateFailed` error if the
+    /// factor source is not found.
+    ///
+    /// # Emits Event
+    /// Emits `Event::ProfileModified { change: EventProfileModified::FactorSourceUpdated { id } }`
+    pub async fn update_factor_source(
+        &self,
+        updated: FactorSource,
+    ) -> Result<()> {
+        let id = updated.factor_source_id();
+
+        debug!("Updating FactorSource with ID: {}", &id);
+
+        self.update_profile_with(|mut p| {
+            // p.update_last_used_of_factor_source(&id)
+            p.update_any_factor_source(&id, |fs| *fs = updated.clone())
+                .map_err(|_| CommonError::UpdateFactorSourceMutateFailed)
+        })
+        .await?;
+
+        self.event_bus
+            .emit(EventNotification::profile_modified(
+                EventProfileModified::FactorSourceUpdated { id },
+            ))
+            .await;
+
+        Ok(())
+    }
+
+    /// Returns `Ok(false)` if the Profile already contained a factor source with the
+    /// same id (Profile unchanged occurred).
+    ///
+    /// # Emits Event
+    /// Emits `Event::ProfileModified { change: EventProfileModified::FactorSourceAdded }`
+    ///
+    /// And also emits `Event::ProfileModified { change: EventProfileModified::FactorSourceUpdated }`,
+    /// if the newly added FactorSource is a new **main** flag, then we remove the
+    /// main flag from the old BDFS.
+    ///
+    /// And also emits `Event::ProfileSaved` after having successfully written the JSON
+    /// of the active profile to secure storage.
+    pub async fn add_factor_source(
+        &self,
+        factor_source: FactorSource,
+    ) -> Result<bool> {
+        let id = factor_source.factor_source_id();
+
+        let contains = self.factor_source_ids().contains(&id);
+
+        if contains {
+            return Ok(false);
+        }
+
+        self.add_factor_sources_without_emitting_factor_sources_added(
+            FactorSources::just(factor_source),
+        )
+        .await?;
+
+        self.event_bus
+            .emit(EventNotification::profile_modified(
+                EventProfileModified::FactorSourceAdded { id },
+            ))
+            .await;
+
+        Ok(true)
+    }
+
+    pub async fn add_factor_sources(
+        &self,
+        factor_sources: FactorSources,
+    ) -> Result<Vec<FactorSourceID>> {
+        let ids = self
+            .add_factor_sources_without_emitting_factor_sources_added(
+                factor_sources,
+            )
+            .await?;
+
+        self.event_bus
+            .emit(EventNotification::profile_modified(
+                EventProfileModified::FactorSourcesAdded { ids: ids.clone() },
+            ))
+            .await;
+
+        Ok(ids)
+    }
+
+    pub async fn debug_add_all_sample_factors(
```

Comment:
Should this be under `#[cfg(test)]` ?

Reply:
nope, it is UniFFI exported.

Reply:
where is that? couldn't find it

Reply:
check the `impl` scope in which this method is defined, it has `#[uniffi::export]`. This us a UniFFI **object** not a record, and for Objects in UniFFI we add the `#[uniffi::export]` attached to the `impl` itself, so all methods inside that impl are exported.